### PR TITLE
Dev/pickle properties

### PR DESCRIPTION
**Pull request by bp-kelley**
Adds optional pickling of properties.

```
namespace PropertyPickleOptions {
  const unsigned int Default = 0; // no data pickled
  const unsigned int AllProps = 0xFFFFFFFF; // all data pickled
  const unsigned int MolProps = 0x1; // only public non computed properties
  const unsigned int AtomProps = 0x10;
  const unsigned int BondProps = 0x100;
  const unsigned int PrivateProps = 0x10000;
  const unsigned int ComputedProps = 0x100000;
}
  
  static void pickleMol(const ROMol *mol, std::ostream &ss,
                        unsigned int propertyFlags=PropertyPickleOptions::Default);
```

Not yet exposed to the python API yet.

Need to see what happens when pickled properties are unpickled from older rdkits, but I think its ok (they are ignored).

### CODE REVIEW

**Comment by bp-kelley** [comment]
@greglandrum The last commit 20ce5ea adds a bit of a hack that fixes the reaction binary.

This allows both the atom and query to be pickled.  This flag is holistic to a molecule (it was the only place the flag has space.)  It is *not* backwards compatible so it may not be acceptable in the current form.

**Comment by bp-kelley** [comment]
@greglandrum reverted the reaction pickler fix, working on a backwards compatible version

**Comment by bp-kelley** [comment]
@greglandrum it doesn't look like I can easily make a backwards compatible version easily as tags cannot be readily ignored (the size isn't stored after the tag, so it can't be skipped).

I suppose that this means if we write one of the new tags, then the version should be bumped but if we don't, the version should be the older version?  Would this work?

**Comment by greglandrum** [comment]
I answered the question above elsewhere, but just to capture it here: we decided to just bump the version all the time.

File: Code/GraphMol/MolPickler.cpp [hunk 1]
```diff
@@ -810,6 +789,41 @@ void MolPickler::_pickle(const ROMol *mol, std::ostream &ss) {
...
+         bondIt++) {
+      _pickleProperties(ss, **bondIt, propertyFlags);
+    }
+    streamWrite(ss, ENDPROPS);
+  }
+
+  if(propertyFlags & PropertyPickleOptions::QueryAtomData) {
+    streamWrite(ss, BEGINQUERYATOMDATA);
+    for (ROMol::ConstAtomIterator atIt = mol->beginAtoms(); atIt != mol->endAtoms();
+         atIt++) {
```
Comment (bp-kelley) [inline Code/GraphMol/MolPickler.cpp]: Ug, I didn't notice this.  Seeing post-increment makes me throw up a little in my mouth.  I think I just copied the snippet from  a few lines below.  I'l fix it there as well.

File: Code/GraphMol/MolPickler.h [hunk 2]
```diff
@@ -42,6 +42,18 @@ class MolPicklerException : public std::exception {
   std::string _msg;
 };
 
+namespace PropertyPickleOptions {
+  const unsigned int Default = 0; // no data pickled
```
Comment (bp-kelley) [inline Code/GraphMol/MolPickler.h]: None isn't easy to wrap in python.  How about NoProperties?

File: Code/GraphMol/ChemReactions/ReactionPickler.cpp [hunk 3]
```diff
@@ -137,7 +137,7 @@ void ReactionPickler::_pickle(const ChemicalReaction *rxn, std::ostream &ss) {
   streamWrite(ss, BEGINPRODUCTS);
   for (MOL_SPTR_VECT::const_iterator tmpl = rxn->beginProductTemplates();
        tmpl != rxn->endProductTemplates(); ++tmpl) {
-    MolPickler::pickleMol(tmpl->get(), ss);
+    MolPickler::pickleMol(tmpl->get(), ss, PropertyPickleOptions::AllProps);
```
Comment (bp-kelley) [inline Code/GraphMol/ChemReactions/ReactionPickler.cpp]: Yes, this fixes the stereo issues in the products.

File: Code/GraphMol/testPickler.cpp [hunk 4]
```diff
@@ -1039,39 +1038,200 @@ void testPickleProps() {
...
+    MolPickler::pickleMol(*m, pkl, PropertyPickleOptions::AllProps);
+    RWMol *m2 = new RWMol(pkl);
+    TEST_ASSERT(m2);
+    TEST_ASSERT(m2->getProp<double>("double") == 1.0);
+    TEST_ASSERT(m2->getProp<int>("int") == 100);
+    TEST_ASSERT(m2->getProp<bool>("bool") == true);
+    TEST_ASSERT(m2->getProp<bool>("boolfalse") == false);
+    //TEST_ASSERT(m->getProp<std::vector<double> >("dvec") == v);
+    
+    BOOST_LOG(rdErrorLog) << "\tatoms" << std::endl;
```
Comment (greglandrum) [inline Code/GraphMol/testPickler.cpp]: Probably not bad to remove these debugging outputs, but that's not a big deal

**Comment by bp-kelley** [comment]
I'll think about that.  Should I put props on conformers while we're at it?

**Comment by greglandrum** [comment]
I'd save the conformer properties for a separate PR

File: Code/GraphMol/testPicklerGlobalSettings.cpp [hunk 5]
```diff
@@ -0,0 +1,239 @@
+// $Id$
```
Comment (greglandrum) [inline Code/GraphMol/testPicklerGlobalSettings.cpp]: This header should be cleaned up: the Id thing is an SVN dropping and the copyright is wrong.

Comment (bp-kelley) [inline Code/GraphMol/testPicklerGlobalSettings.cpp]: I'll do that this morning.  That was a copy paste of course.

**Comment by greglandrum** [comment]
sorry, I spaced on doing this merge.

**Pull request merged**